### PR TITLE
More libgit2 docs

### PR DESCRIPTION
**Pull request by kshyatt**

### CODE REVIEW

File: base/libgit2/libgit2.jl [hunk 1]
```diff
@@ -42,22 +42,37 @@ immutable State
...
 
-"""Return HEAD GitHash as string"""
+"""
+    head(pkg::AbstractString) -> String
+
+Return current HEAD [`GitHash`](@ref) of
+the `pkg` repo as a string.
+"""
 function head(pkg::AbstractString)
     with(GitRepo, pkg) do repo
```
Comment (KristofferC) [inline base/libgit2/libgit2.jl]: Just checking, could this `with` be removed now?

File: base/libgit2/libgit2.jl [hunk 2]
```diff
@@ -74,18 +89,18 @@ function iscommit(id::AbstractString, repo::GitRepo)
 end
 
 """
-    LibGit2.isdirty(repo::GitRepo[, paths]; cached=false)
+    LibGit2.isdirty(repo::GitRepo, paths::AbstractString=""; cached::Bool=false) -> Bool
```
Comment (KristofferC) [inline base/libgit2/libgit2.jl]: What is `paths` here?

File: base/libgit2/libgit2.jl [hunk 3]
```diff
@@ -74,18 +89,18 @@ function iscommit(id::AbstractString, repo::GitRepo)
...
 
-See `git diff-index HEAD [-- <path>]`
+See `git diff-index HEAD [-- <path>]`.
 """
 isdirty(repo::GitRepo, paths::AbstractString=""; cached::Bool=false) =
     isdiff(repo, Consts.HEAD_FILE, paths, cached=cached)
 
 """
-    LibGit2.isdiff(repo::GitRepo, treeish[, paths]; cached=false)
+    LibGit2.isdiff(repo::GitRepo, treeish::AbstractString, paths::AbstractString=""; cached::Bool=false)
```
Comment (KristofferC) [inline base/libgit2/libgit2.jl]: Same here about paths

File: base/libgit2/libgit2.jl (hunk 1 above)
Comment (kshyatt) [inline base/libgit2/libgit2.jl]: I have another PR for that :)

File: base/libgit2/libgit2.jl (hunk 2 above)
Comment (kshyatt) [inline base/libgit2/libgit2.jl]: Oooo, good catch.

File: base/libgit2/libgit2.jl [hunk 4]
```diff
@@ -181,7 +223,16 @@ function fetch{T<:AbstractString, P<:AbstractCredentials}(repo::GitRepo;
     end
 end
 
-""" git push [<url>|<repository>] [<refspecs>]"""
+"""
+    push(repo::GitRepo, remote::AbstractString="origin", remoteurl::AbstractString="", refspecs=AbstractString[], force::Bool=false, payload=Nullable{AbstractCredentials}())
+
+Equivalent to `git push [<url>|<repository>] [<refspecs>]`.
```
Comment (KristofferC) [inline base/libgit2/libgit2.jl]: `url` and `repository` are not a part of the function arguments.

File: base/libgit2/libgit2.jl [hunk 5]
```diff
@@ -161,7 +195,15 @@ function make_payload{P<:AbstractCredentials}(payload::Nullable{P})
     Ref{Nullable{AbstractCredentials}}(payload)
 end
 
-""" git fetch [<url>|<repository>] [<refspecs>]"""
+"""
+    fetch(repo::GitRepo, remote::AbstractString="origin", remoteurl::AbstractString="", refspecs=AbstractString[], payload=Nullable{AbstractCredentials}())
+
+Equivalent to `git fetch [<url>|<repository>] [<refspecs>]`.
+Fetches updates from the `remote` upstream of `repo`, looking at `remoteurl`.
```
Comment (ararslan) [inline base/libgit2/libgit2.jl]: Might be nice as "Fetches updates from the upstream `remote` with URL `remoteurl` of the repository `repo`."

File: base/libgit2/libgit2.jl [hunk 6]
```diff
@@ -212,7 +268,16 @@ function branch(repo::GitRepo)
     end
 end
 
-""" git checkout [-b|-B] <branch> [<start-point>] [--track <remote>/<branch>] """
+"""
+    branch!(repo::GitRepo, branch_name::AbstractString, commit::AbstractString="", track::AbstractString="", force::Bool=false, set_head::Bool=true)
+
+Equivalent to `git checkout [-b|-B] <branch> [<start-point>] [--track <remote>/<branch>]`.
```
Comment (ararslan) [inline base/libgit2/libgit2.jl]: Would be nice to map the function arguments to the appropriate pieces in this Git command

**Comment by StefanKarpinski** [comment]
@KristofferC: could you leave a review requesting changes – makes it easier to keep track of the fact that this isn't (apparently) quite ready to merge just yet.

**Comment by kshyatt** [comment]
Updated to address comments.

**Comment by kshyatt** [comment]
Ready to go?

File: base/libgit2/libgit2.jl [hunk 7]
```diff
@@ -74,21 +89,23 @@ function iscommit(id::AbstractString, repo::GitRepo)
 end
 
 """
-    LibGit2.isdirty(repo::GitRepo[, paths]; cached=false)
+    LibGit2.isdirty(repo::GitRepo, paths::AbstractString=""; cached::Bool=false) -> Bool
 
 Checks if there have been any changes to tracked files in the working tree (if
 `cached=false`) or the index (if `cached=true`).
+`paths` are the specifications for options for the diff.
```
Comment (tkelman) [inline base/libgit2/libgit2.jl]: odd name for that variable then?

File: base/libgit2/libgit2.jl [hunk 8]
```diff
@@ -74,21 +89,23 @@ function iscommit(id::AbstractString, repo::GitRepo)
...
 
-See `git diff-index HEAD [-- <path>]`
+See `git diff-index HEAD [-- <path>]`.
```
Comment (tkelman) [inline base/libgit2/libgit2.jl]: "equivalent to" ?

File: base/libgit2/libgit2.jl [hunk 9]
```diff
@@ -132,11 +149,24 @@ function diff_files(repo::GitRepo, branch1::AbstractString, branch2::AbstractStr
     return files
 end
 
+"""
+    is_ancestor_of(a::AbstractString, b::AbstractString, repo::GitRepo) -> Bool
+
+Returns `true` if `a`, a [`GitHash`](@ref) in string form, is an ancestor of
+`b`, a [`GitHash`](@ref) in string form. `a` and `b` could be branches, tags,
```
Comment (tkelman) [inline base/libgit2/libgit2.jl]: if they're branches or tags then they aren't really `GitHash`es

File: base/libgit2/libgit2.jl [hunk 10]
```diff
@@ -181,7 +225,16 @@ function fetch{T<:AbstractString, P<:AbstractCredentials}(repo::GitRepo;
     end
 end
 
-""" git push [<url>|<repository>] [<refspecs>]"""
+"""
+    push(repo::GitRepo, remote::AbstractString="origin", remoteurl::AbstractString="", refspecs=AbstractString[], force::Bool=false, payload=Nullable{AbstractCredentials}())
+
+Equivalent to `git push [<remoteurl>|<repo>] [<refspecs>]`.
+Pushes updates to the `remote` upstream of `repo`, looking at `remoteurl`.
```
Comment (tkelman) [inline base/libgit2/libgit2.jl]: what does "looking at" mean here? won't most remotes have a url already?

File: base/libgit2/libgit2.jl [hunk 11]
```diff
@@ -161,7 +197,15 @@ function make_payload{P<:AbstractCredentials}(payload::Nullable{P})
     Ref{Nullable{AbstractCredentials}}(payload)
 end
 
-""" git fetch [<url>|<repository>] [<refspecs>]"""
+"""
+    fetch(repo::GitRepo, remote::AbstractString="origin", remoteurl::AbstractString="", refspecs=AbstractString[], payload=Nullable{AbstractCredentials}())
```
Comment (tkelman) [inline base/libgit2/libgit2.jl]: everything but `repo` is a kwarg

File: base/libgit2/libgit2.jl [hunk 12]
```diff
@@ -181,7 +225,16 @@ function fetch{T<:AbstractString, P<:AbstractCredentials}(repo::GitRepo;
     end
 end
 
-""" git push [<url>|<repository>] [<refspecs>]"""
+"""
+    push(repo::GitRepo, remote::AbstractString="origin", remoteurl::AbstractString="", refspecs=AbstractString[], force::Bool=false, payload=Nullable{AbstractCredentials}())
```
Comment (tkelman) [inline base/libgit2/libgit2.jl]: everything but `repo` is a kwarg

File: base/libgit2/libgit2.jl [hunk 13]
```diff
@@ -212,7 +270,16 @@ function branch(repo::GitRepo)
     end
 end
 
-""" git checkout [-b|-B] <branch> [<start-point>] [--track <remote>/<branch>] """
+"""
+    branch!(repo::GitRepo, branch_name::AbstractString, commit::AbstractString="", track::AbstractString="", force::Bool=false, set_head::Bool=true)
```
Comment (tkelman) [inline base/libgit2/libgit2.jl]: `track`, `force`, and `set_head` are kwargs

File: base/libgit2/libgit2.jl [hunk 14]
```diff
@@ -212,7 +270,16 @@ function branch(repo::GitRepo)
...
+
+Equivalent to `git checkout [-b|-B] <branch_name> [<commit>] [--track <track>]`.
+Checkout a new git branch in the `repo` repository. `commit` is the [`GitHash`](@ref),
+in string form, which will be the start of the new branch. `track` is the name of the
+remote branch this new branch should track, if any. If empty (the default), no remote branch
+will be tracked. If `force` is `true`, branch creation will be forced. If `set_head` is
+`true`, as the branch creation finishes the branch head will be set as the HEAD of `repo`.
```
Comment (tkelman) [inline base/libgit2/libgit2.jl]: after the branch creation finishes? doesn't make much sense for it to happen at exactly the same time

File: base/libgit2/libgit2.jl (hunk 7 above)
Comment (kshyatt) [inline base/libgit2/libgit2.jl]: I'll call it `pathspecs`?

File: base/libgit2/libgit2.jl (hunk 9 above)
Comment (kshyatt) [inline base/libgit2/libgit2.jl]: I'm pretty sure the args here must be string-ified Hashes which are OIDs for the branch or tag.

Comment (tkelman) [inline base/libgit2/libgit2.jl]: right by saying GitHash implies the julia type

Comment (tkelman) [inline base/libgit2/libgit2.jl]: a branch or a tag sounds like you'd refer to it by name, since a hash doesn't care where it's from

Comment (kshyatt) [inline base/libgit2/libgit2.jl]:
```julia
function merge_base(repo::GitRepo, one::AbstractString, two::AbstractString)
    oid1_ptr = Ref(GitHash(one))
    oid2_ptr = Ref(GitHash(two))
    moid_ptr = Ref(GitHash())
```

Comment (kshyatt) [inline base/libgit2/libgit2.jl]: You do not refer to it by name, no matter if it's a commit/branch/tag. `a` can be a commit and `b` can be a tag, even.

Comment (tkelman) [inline base/libgit2/libgit2.jl]: I woukd just leave off the "can be branches tags" etc sentence then. I wouldn't call a sha a branch usually - it kind of works but it's not what you usually refer to them by.

Comment (kshyatt) [inline base/libgit2/libgit2.jl]: How about `a` and `b` are SHAs which may refer to commits/tags/branches?

Comment (kshyatt) [inline base/libgit2/libgit2.jl]: I guess I want to make it clear that you can use this function to see if `a` is a derived branch of `b` or whatever.

Comment (tkelman) [inline base/libgit2/libgit2.jl]: I think "ancestor" makes that clear enough without having to mention anything about branches or tags.

File: base/libgit2/libgit2.jl [hunk 15]
```diff
@@ -212,10 +270,19 @@ function branch(repo::GitRepo)
...
+Checkout a new git branch in the `repo` repository. `commit` is the [`GitHash`](@ref),
+in string form, which will be the start of the new branch. `track` is the name of the
+remote branch this new branch should track, if any. If empty (the default), no remote branch
+will be tracked. If `force` is `true`, branch creation will be forced. If `set_head` is
+`true`, after the branch creation finishes the branch head will be set as the HEAD of `repo`.
+"""
 function branch!(repo::GitRepo, branch_name::AbstractString,
                  commit::AbstractString = ""; # start point
-                 track::AbstractString  = "", # track remote branch
+                 track::AbstractString  = "", #track remote branch
```
Comment (tkelman) [inline base/libgit2/libgit2.jl]: this change seems unnecessary?

Comment (kshyatt) [inline base/libgit2/libgit2.jl]: I don't remember doing that...

Comment (tkelman) [inline base/libgit2/libgit2.jl]: haunted editor

File: base/libgit2/libgit2.jl (hunk 9 above)
Comment (kshyatt) [inline base/libgit2/libgit2.jl]: Fixed just now.

File: base/libgit2/libgit2.jl (hunk 15 above)
Comment (kshyatt) [inline base/libgit2/libgit2.jl]: Fixed.

File: base/libgit2/libgit2.jl [hunk 16]
```diff
@@ -161,7 +196,15 @@ function make_payload{P<:AbstractCredentials}(payload::Nullable{P})
     Ref{Nullable{AbstractCredentials}}(payload)
 end
 
-""" git fetch [<url>|<repository>] [<refspecs>]"""
+"""
+    fetch(repo::GitRepo; remote::AbstractString="origin", remoteurl::AbstractString="", refspecs=AbstractString[], payload=Nullable{AbstractCredentials}())
+
+Equivalent to `git fetch [<remoteurl>|<repo>] [<refspecs>]`.
+Fetches updates from the upstream `remote` with URL `remoteurl` of the repository repo.
```
Comment (tkelman) [inline base/libgit2/libgit2.jl]: repo should be code highlighted?

File: base/libgit2/libgit2.jl [hunk 17]
```diff
@@ -132,11 +149,23 @@ function diff_files(repo::GitRepo, branch1::AbstractString, branch2::AbstractStr
     return files
 end
 
+"""
+    is_ancestor_of(a::AbstractString, b::AbstractString, repo::GitRepo) -> Bool
+
+Returns `true` if `a`, a [`GitHash`](@ref) in string form, is an ancestor of
+`b`, a [`GitHash`](@ref) in string form.
```
Comment (ararslan) [inline base/libgit2/libgit2.jl]: Probably an unnecessary change, but I think this would read slightly better as
> Return `true` if `a` is an ancestor of `b`, where `a` and `b` are ``[`GitHash`](@ref)``es in string form.

**Comment by kshyatt** [comment]
@tkelman is this ready to roll?

**Comment by kshyatt** [comment]
OK, well, I addressed all the comments and gave everyone a day. If there's something more to be done I'll open another PR.

**Comment by tkelman** [comment]
didn't address @ararslan's most recent comment

**Comment by ararslan** [comment]
I assumed that she considered it an unnecessary change. ¯\\\_(ツ)\_/¯

File: base/libgit2/libgit2.jl [hunk 18]
```diff
@@ -42,22 +42,37 @@ immutable State
     work::GitHash
 end
 
-"""Return HEAD GitHash as string"""
+"""
+    head(pkg::AbstractString) -> String
```
Comment (tkelman) [inline base/libgit2/libgit2.jl]: strange for this to be using `pkg` as an argument name given it's more general than that

(though most places that use `repo` do it for a `::GitRepo` input)
